### ISSUE TITLE
Investigate the test_module_utils_basic__log_invocation failures

### ISSUE THREAD

**Issue by sivel** [ansible/ansible#15108]
We had an issue with run_tests.sh that was masking some unit test errors.  This was fixed in https://github.com/ansible/ansible/pull/15105

However, I was unable to determine the cause to why `test_module_utils_basic__log_invocation` was failing, and at the recommendation of @abadger I have temporarily skipped this test.

It seems that the order in which the `Invoked with` message is built changes between runs.  I don't believe this was always the case, so we should figure out why.

I had originally implemented a fix where `log_args` was `sorted`, which would produce a predictable order, although I am unsure if that is the right fix.  It may actually be.

Additionally, this test should be skipped for py3 once we figure out why the test was failing.

```
======================================================================
FAIL: test_module_utils_basic__log_invocation (units.module_utils.basic.test__log_invocation.TestModuleUtilsBasic)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/ansible/ansible/test/units/module_utils/basic/test__log_invocation.py", line 56, in test_module_utils_basic__log_invocation
    'no_log': 'NOT_LOGGING_PARAMETER',
  File "/home/travis/build/ansible/ansible/.tox/py27/lib/python2.7/site-packages/mock.py", line 835, in assert_called_with
    raise AssertionError(msg)
AssertionError: Expected call: mock('Invoked with bam=bam bar=[1, 2, 3] foo=False baz=baz no_log=NOT_LOGGING_PARAMETER password=NOT_LOGGING_PASSWORD ', log_args={'foo': 'False', 'password': 'NOT_LOGGING_PASSWORD', 'bam': 'bam', 'bar': '[1, 2, 3]', 'no_log': 'NOT_LOGGING_PARAMETER', 'baz': 'baz'})
Actual call: mock('Invoked with foo=False password=NOT_LOGGING_PASSWORD bam=bam bar=[1, 2, 3] no_log=NOT_LOGGING_PARAMETER baz=baz ', log_args={'foo': 'False', 'password': 'NOT_LOGGING_PASSWORD', 'bam': 'bam', 'bar': '[1, 2, 3]', 'no_log': 'NOT_LOGGING_PARAMETER', 'baz': 'baz'})
    "Expected call: mock('Invoked with bam=bam bar=[1, 2, 3] foo=False baz=baz no_log=NOT_LOGGING_PARAMETER password=NOT_LOGGING_PASSWORD ', log_args={'foo': 'False', 'password': 'NOT_LOGGING_PASSWORD', 'bam': 'bam', 'bar': '[1, 2, 3]', 'no_log': 'NOT_LOGGING_PARAMETER', 'baz': 'baz'})\nActual call: mock('Invoked with foo=False password=NOT_LOGGING_PASSWORD bam=bam bar=[1, 2, 3] no_log=NOT_LOGGING_PARAMETER baz=baz ', log_args={'foo': 'False', 'password': 'NOT_LOGGING_PASSWORD', 'bam': 'bam', 'bar': '[1, 2, 3]', 'no_log': 'NOT_LOGGING_PARAMETER', 'baz': 'baz'})" = self._format_mock_failure_message(('Invoked with bam=bam bar=[1, 2, 3] foo=False baz=baz no_log=NOT_LOGGING_PARAMETER password=NOT_LOGGING_PASSWORD ',), {'log_args': {'foo': 'False', 'password': 'NOT_LOGGING_PASSWORD', 'bam': 'bam', 'bar': '[1, 2, 3]', 'no_log': 'NOT_LOGGING_PARAMETER', 'baz': 'baz'}})
>>  raise AssertionError("Expected call: mock('Invoked with bam=bam bar=[1, 2, 3] foo=False baz=baz no_log=NOT_LOGGING_PARAMETER password=NOT_LOGGING_PASSWORD ', log_args={'foo': 'False', 'password': 'NOT_LOGGING_PASSWORD', 'bam': 'bam', 'bar': '[1, 2, 3]', 'no_log': 'NOT_LOGGING_PARAMETER', 'baz': 'baz'})\nActual call: mock('Invoked with foo=False password=NOT_LOGGING_PASSWORD bam=bam bar=[1, 2, 3] no_log=NOT_LOGGING_PARAMETER baz=baz ', log_args={'foo': 'False', 'password': 'NOT_LOGGING_PASSWORD', 'bam': 'bam', 'bar': '[1, 2, 3]', 'no_log': 'NOT_LOGGING_PARAMETER', 'baz': 'baz'})")
```

**Comment by jimi-c** [comment]
I have been trying to avoid making new tests skip for py3, unless we have a really good reason. Is this failing on devel or stable-2.0? I had disabled it on 2.0 but it was working fine on devel.

**Comment by sivel** [comment]
This was on devel.  On py2.x the failure was due to the order of args changing in each `Invoked with` message, so the `assert_called_with` never matched.

On py3 this was failing for a completely different reason in that it was saying `baz` was an invalid argument for the module.

You can see both sets of failures on the following build:

https://travis-ci.org/ansible/ansible/builds/117867179

This was exposed after fixing the syntax errors in the ovirt inventory and adding `set -e` globally in run_tests.sh

**Comment by abadger** [comment]
I'm not sure how this ever worked on python3.3+ because dictionaries and other hashables get a random seed added with every new interpreter run.  But it is passing locally (python-3.4.4).  On python-2.7 (and rhel's 2.6 as a security backport), this is controlled by an environment variable PYTHONHASHSEED.

We could explicitly sort the params before constructing the message to make this more testable (but there's nothing wrong with the current code.  It's just more testable if the order is defined).  We could also look at testing the string message separately from the dictionary.

**Comment by sivel** [comment]
It was the "not sure how this ever worked" that made me pause and not include my fix.

The fix was using `sorted(log_args)` in basic.py and updating the expected string in the tests to match.

**Comment by abadger** [comment]
Fixed for python2 here: d3583108eca98f72b9b4898a5cc5e9cf1cacf251  Updated the test to check the message in pieces rather than all at once.

Python3 actually has deeper issues because we're transforming all the parameters to bytes as that's what we've always sent on python2 (so it's backwards compatibility).  For Python3, we may want to revamp this so that the parameters are interpreted as text type instead.  But if we do that then we have to audit and check a whole slew of code in basic.py to make sure that it is doing the right thing.  Going to punt on that for now and simply skip the test on python3.
